bool bool_data
int32 int_data